quark.c

```diff
@@ -4736,6 +4738,217 @@ raw_event_tty(struct quark_queue *qq, struct raw_event *raw)
+		    qev->process->filename != NULL)
+			return (path_match(field, qev->process->filename));
+		break;
+	case RF_FILE_PATH:
+		if (qev->file != NULL)
+			return (path_match(field, qev->file->path));
+		break;
+	default:
+		break;
+	}
+
+	return (0);
+}
+
+struct quark_rule *
+quark_ruleset_match(struct quark_ruleset *ruleset, struct quark_event *qev)
+{
+	struct quark_rule	*rule;
+	struct rule_field	*field;
+	size_t			 i, j;
+
+	for (i = 0; i < ruleset->n_rules; i++) {
+		rule = ruleset->rules + i;
+		rule->evals++;
+		for (j = 0; j < rule->n_fields; j++) {
+			field = rule->fields + j;
+			if (!rule_field_match(rule, field, qev))
+				break;
+		}
+		/* All fields matched, this is the rule! */
+		if (j == rule->n_fields) {
+			rule->hits++;
+			return (rule);
+		}
+	}
+
+	return (NULL);
+}
+
+struct quark_rule *
+quark_ruleset_append_rule(struct quark_ruleset *ruleset, int action)
+{
+	struct quark_rule	*new_rules, *rule;
+	size_t			 new_n_rules;
+
+	if (action != RA_DROP && action != RA_PASS)
+		return (errno = EINVAL, NULL);
+	new_n_rules = ruleset->n_rules + 1;
+	new_rules = reallocarray(ruleset->rules, new_n_rules,
+	    sizeof(*ruleset->rules));
+	if (new_rules == NULL)
+		return (NULL);
+
+	ruleset->rules = new_rules;
+	ruleset->n_rules = new_n_rules;
+
+	rule = &ruleset->rules[ruleset->n_rules - 1];
+	rule->fields = NULL;
+	rule->n_fields = 0;
+	rule->action = action;
+	rule->evals = 0;
+	rule->hits = 0;
+
+	return (rule);
+}
+
+static int
+quark_rule_append_field(struct quark_rule *rule, struct rule_field *rf)
+{
+	struct rule_field	*new_fields;
+	size_t			 new_n_fields;
+
+	switch (rf->code) {
+	case RF_PROCESS_PID:		/* FALLTHROUGH */
+	case RF_PROCESS_PPID:
+		if (rf->pid == 0)
+			goto inval;
+		break;
+	case RF_PROCESS_FILENAME:	/* FALLTHROUGH */
+	case RF_FILE_PATH:
+		if (*rf->path == 0)
+			goto inval;
+		if (rf->path[sizeof(rf->path) - 1] != 0)
+			goto inval;
+		break;
+	default:
+		goto inval;
+	}
+
+	new_n_fields = rule->n_fields + 1;
+	new_fields = reallocarray(rule->fields, new_n_fields,
+	    sizeof(*rule->fields));
+	if (new_fields == NULL)
+		return (-1);
+
+	rule->fields = new_fields;
+	rule->n_fields = new_n_fields;
+	rule->fields[rule->n_fields - 1] = *rf;
+
+	return (0);
+
+inval:
+	errno = EINVAL;
+	return (-1);
+}
+
+static int
+quark_rule_append_any_pid(struct quark_rule *rule, u32 pid,
+    enum rule_field_code code)
+{
+	struct rule_field	rf;
+
+	bzero(&rf, sizeof(rf));
+	rf.code = code;
+	rf.pid = pid;
+
+	return (quark_rule_append_field(rule, &rf));
+}
+
+int
+quark_rule_append_pid(struct quark_rule *rule, u32 pid)
+{
+	return (quark_rule_append_any_pid(rule, pid, RF_PROCESS_PID));
+}
+
+int
+quark_rule_append_ppid(struct quark_rule *rule, u32 pid)
+{
+	return (quark_rule_append_any_pid(rule, pid, RF_PROCESS_PPID));
+}
+
+static int
+quark_rule_append_any_path(struct quark_rule *rule, const char *path,
+    enum rule_field_code code)
+{
+	struct rule_field	 rf;
+	char			*p;
+
+	if (strlen(path) >= PATH_MAX)
```

Comment:
should evaluate against `sizeof(rf.path)` in case we make it smaller, which we probably won't, but still.

Reply:
I did this intentionally cause I want to make rf.path a pointer and allocate it, so I'd leave it like this so I don't have to remember to change it.